File: android-compose/app/src/main/java/com/ohmz/tday/compose/core/data/list/ListRepository.kt
package com.ohmz.tday.compose.core.data.list

import android.util.Log
import com.ohmz.tday.compose.core.data.CachedListRecord
import com.ohmz.tday.compose.core.data.MutationKind
import com.ohmz.tday.compose.core.data.OfflineSyncState
import com.ohmz.tday.compose.core.data.PendingMutationRecord
import com.ohmz.tday.compose.core.data.SecureConfigStore
import com.ohmz.tday.compose.core.data.cache.LOCAL_LIST_PREFIX
import com.ohmz.tday.compose.core.data.cache.OfflineCacheManager
import com.ohmz.tday.compose.core.data.cache.listFromCache
import com.ohmz.tday.compose.core.data.cache.orderListsLikeWeb
import com.ohmz.tday.compose.core.data.cache.parseOptionalInstant
import com.ohmz.tday.compose.core.data.isLikelyUnrecoverableMutationError
import com.ohmz.tday.compose.core.data.requireApiBody
import com.ohmz.tday.compose.core.data.sync.SyncManager
import com.ohmz.tday.compose.core.model.CreateListRequest
import com.ohmz.tday.compose.core.model.DeleteListRequest
import com.ohmz.tday.compose.core.model.ListSummary
import com.ohmz.tday.compose.core.model.UpdateListRequest
import com.ohmz.tday.compose.core.model.capitalizeFirstListLetter
import com.ohmz.tday.compose.core.network.TdayApiService
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ListRepository @Inject constructor(
    private val api: TdayApiService,
    private val cacheManager: OfflineCacheManager,
    private val secureConfigStore: SecureConfigStore,
    private val syncManager: SyncManager,
) {
    suspend fun fetchLists(): List<ListSummary> {
        return buildListsForState(cacheManager.loadOfflineState())
    }

    fun fetchListsSnapshot(): List<ListSummary> {
        return buildListsForState(cacheManager.loadOfflineStateBlocking())
    }

    suspend fun createList(
        name: String,
        color: String? = null,
        iconKey: String? = null,
    ) {
        val normalizedName = capitalizeFirstListLetter(name).trim()
        if (normalizedName.isBlank()) return

        val localListId = "$LOCAL_LIST_PREFIX${UUID.randomUUID()}"
        val timestampMs = System.currentTimeMillis()
        val mutationId = UUID.randomUUID().toString()
        cacheManager.updateOfflineState { state ->
            val newList = CachedListRecord(
                id = localListId,
                name = normalizedName,
                color = color,
                iconKey = iconKey,
                todoCount = 0,
                createdAtEpochMs = timestampMs,
                updatedAtEpochMs = timestampMs,
            )
            state.copy(
                lists = state.lists + newList,
                pendingMutations = state.pendingMutations + PendingMutationRecord(
                    mutationId = mutationId,
                    kind = MutationKind.CREATE_LIST,
                    targetId = localListId,
                    timestampEpochMs = timestampMs,
                    name = normalizedName,
                    color = color,
                    iconKey = iconKey,
                ),
            )
        }

        if (syncManager.isLocalMode()) return

        runCatching {
            requireApiBody(
                api.createList(
                    CreateListRequest(
                        name = normalizedName,
                        color = color,
                        iconKey = iconKey,
                    ),
                ),
                "Could not create list",
            ).list
        }.onSuccess { createdList ->
            if (createdList == null) return@onSuccess
            val createdAt =
                parseOptionalInstant(createdList.createdAt)?.toEpochMilli() ?: timestampMs
            val updatedAt =
                parseOptionalInstant(createdList.updatedAt)?.toEpochMilli() ?: timestampMs
            cacheManager.updateOfflineState { state ->
                val remapped = replaceLocalListId(
                    state = state,
                    localListId = localListId,
                    serverListId = createdList.id,
                )
                val todoCount = remapped.todos.count { !it.completed && it.listId == createdList.id }
                remapped.copy(
                    lists = remapped.lists.map { list ->
                        if (list.id == createdList.id) {
                            list.copy(
                                name = createdList.name,
                                color = createdList.color,
                                iconKey = createdList.iconKey ?: list.iconKey,
                                todoCount = todoCount,
                                updatedAtEpochMs = updatedAt,
                                createdAtEpochMs = createdAt,
                            )
                        } else {
                            list
                        }
                    },
                    pendingMutations = remapped.pendingMutations.filterNot { it.mutationId == mutationId },
                )
            }
        }
    }

    suspend fun updateList(
        listId: String,
        name: String,
        color: String? = null,
        iconKey: String? = null,
    ) {
        val trimmedName = capitalizeFirstListLetter(name).trim()
        if (listId.isBlank()) return
        require(trimmedName.isNotBlank()) { "List name is required" }
        Log.d(LOG_TAG, "updateList start hasColor=${color != null} hasIcon=${iconKey != null}")

        val timestampMs = System.currentTimeMillis()
        val mutationId = UUID.randomUUID().toString()
        if (listId.startsWith(LOCAL_LIST_PREFIX)) {
            cacheManager.updateOfflineState { state ->
                state.copy(
                    lists = state.lists.map { list ->
                        if (list.id == listId) {
                            list.copy(
                                name = trimmedName,
                                color = color ?: list.color,
                                iconKey = iconKey ?: list.iconKey,
                                updatedAtEpochMs = timestampMs,
                            )
                        } else {
                            list
                        }
                    },
                    pendingMutations = state.pendingMutations.map { mutation ->
                        if (mutation.kind == MutationKind.CREATE_LIST && mutation.targetId == listId) {
                            mutation.copy(
                                name = trimmedName,
                                color = color ?: mutation.color,
                                iconKey = iconKey ?: mutation.iconKey,
                                timestampEpochMs = timestampMs,
                            )
                        } else {
                            mutation
                        }
                    },
                )
            }
            iconKey?.takeIf { it.isNotBlank() }?.let {
                secureConfigStore.saveListIcon(listId, it)
            }
            if (syncManager.isLocalMode()) return
            syncManager.syncCachedData(force = true, replayPendingMutations = true)
            Log.d(LOG_TAG, "updateList local-list path finished")
            return
        }

        cacheManager.updateOfflineState { state ->
            state.copy(
                lists = state.lists.map { list ->
                    if (list.id == listId) {
                        list.copy(
                            name = trimmedName,
                            color = color ?: list.color,
                            iconKey = iconKey ?: list.iconKey,
                            updatedAtEpochMs = timestampMs,
                        )
                    } else {
                        list
                    }
                },
                pendingMutations = state.pendingMutations
                    .filterNot { it.kind == MutationKind.UPDATE_LIST && it.targetId == listId } +
                    PendingMutationRecord(
                        mutationId = mutationId,
                        kind = MutationKind.UPDATE_LIST,
                        targetId = listId,
                        timestampEpochMs = timestampMs,
                        name = trimmedName,
                        color = color,
                        iconKey = iconKey,
                    ),
            )
        }

        if (syncManager.isLocalMode()) {
            iconKey?.takeIf { it.isNotBlank() }?.let {
                secureConfigStore.saveListIcon(listId, it)
            }
            return
        }

        Log.d(LOG_TAG, "updateList patch /api/list")
        val pendingMutation = PendingMutationRecord(
            mutationId = mutationId,
            kind = MutationKind.UPDATE_LIST,
            targetId = listId,
            timestampEpochMs = timestampMs,
            name = trimmedName,
            color = color,
            iconKey = iconKey,
        )
        val immediateError = runCatching {
            requireApiBody(
                api.patchListByBody(
                    UpdateListRequest(
                        id = listId,
                        name = trimmedName,
                        color = color,
                        iconKey = iconKey,
                    ),
                ),
                "Could not update list",
            )
        }.exceptionOrNull()

        if (immediateError != null && isLikelyUnrecoverableMutationError(immediateError, pendingMutation)) {
            throw immediateError
        }

        iconKey?.takeIf { it.isNotBlank() }?.let {
            secureConfigStore.saveListIcon(listId, it)
        }

        if (immediateError == null) {
            cacheManager.updateOfflineState { state ->
                state.copy(
                    pendingMutations = state.pendingMutations.filterNot { it.mutationId == mutationId },
                )
            }
            Log.d(LOG_TAG, "updateList success")
        } else {
            Log.w(LOG_TAG, "updateList deferred reason=${immediateError.javaClass.simpleName}")
        }
    }

    suspend fun deleteList(
        listId: String,
        onOptimisticDelete: () -> Unit = {},
    ) {
        val normalizedListId = listId.trim()
        if (normalizedListId.isBlank()) return

        val timestampMs = System.currentTimeMillis()
        val mutationId = UUID.randomUUID().toString()
        val pendingMutation = PendingMutationRecord(
            mutationId = mutationId,
            kind = MutationKind.DELETE_LIST,
            targetId = normalizedListId,
            timestampEpochMs = timestampMs,
        )
        val isLocalOnly = normalizedListId.startsWith(LOCAL_LIST_PREFIX)

        cacheManager.updateOfflineState { state ->
            val deletedTodoIds = state.todos
                .filter { it.listId == normalizedListId }
                .map { it.canonicalId }
                .toSet()
            val prunedMutations = state.pendingMutations.filterNot { mutation ->
                mutation.targetId == normalizedListId ||
                        mutation.listId == normalizedListId ||
                        deletedTodoIds.contains(mutation.targetId)
            }

            state.copy(
                lists = state.lists.filterNot { it.id == normalizedListId },
                todos = state.todos.filterNot { it.listId == normalizedListId },
                completedItems = state.completedItems.filterNot { completed ->
                    completed.listId == normalizedListId ||
                            completed.originalTodoId?.let(deletedTodoIds::contains) == true
                },
                pendingMutations = if (isLocalOnly) {
                    prunedMutations
                } else {
                    prunedMutations + pendingMutation
                },
            )
        }

        onOptimisticDelete()

        if (syncManager.isLocalMode()) return

        if (isLocalOnly) {
            syncManager.syncCachedData(force = true, replayPendingMutations = true)
            return
        }

        val immediateError = runCatching {
            requireApiBody(
                api.deleteListByBody(DeleteListRequest(id = normalizedListId)),
                "Could not delete list",
            )
        }.exceptionOrNull()

        if (immediateError != null && isLikelyUnrecoverableMutationError(
                immediateError,
                pendingMutation
            )
        ) {
            throw immediateError
        }

        if (immediateError == null) {
            cacheManager.updateOfflineState { state ->
                state.copy(
                    pendingMutations = state.pendingMutations.filterNot { it.mutationId == mutationId },
                )
            }
            Log.d(LOG_TAG, "deleteList success")
        } else {
            Log.w(
                LOG_TAG,
                "deleteList deferred reason=${immediateError.javaClass.simpleName}"
            )
        }
    }

    private fun buildListsForState(state: OfflineSyncState): List<ListSummary> {
        val todoCountsByList = state.todos
            .asSequence()
            .filterNot { it.completed }
            .groupingBy { it.listId }
            .eachCount()
        return orderListsLikeWeb(state.lists).map {
            listFromCache(cache = it, todoCountOverride = todoCountsByList[it.id] ?: 0)
        }
    }

    private fun replaceLocalListId(
        state: OfflineSyncState,
        localListId: String,
        serverListId: String,
    ): OfflineSyncState {
        return state.copy(
            lists = state.lists.map {
                if (it.id == localListId) it.copy(id = serverListId) else it
            },
            todos = state.todos.map {
                if (it.listId == localListId) it.copy(listId = serverListId) else it
            },
            completedItems = state.completedItems.map {
                if (it.listId == localListId) it.copy(listId = serverListId) else it
            },
            pendingMutations = state.pendingMutations.map {
                it.copy(
                    targetId = if (it.targetId == localListId) serverListId else it.targetId,
                    listId = if (it.listId == localListId) serverListId else it.listId,
                )
            },
        )
    }

    private companion object {
        const val LOG_TAG = "ListRepository"
    }
}

File: ios-swiftUI/Tday/Core/Data/List/ListRepository.swift
import Foundation

@MainActor
final class ListRepository {
    private let api: TdayAPIService
    private let cacheManager: OfflineCacheManager
    private let syncManager: SyncManager

    init(api: TdayAPIService, cacheManager: OfflineCacheManager, syncManager: SyncManager) {
        self.api = api
        self.cacheManager = cacheManager
        self.syncManager = syncManager
    }

    func fetchLists() -> [ListSummary] {
        buildLists(from: cacheManager.loadOfflineState())
    }

    func fetchListsSnapshot() -> [ListSummary] {
        buildLists(from: cacheManager.loadOfflineState())
    }

    func createList(name: String, color: String? = nil, iconKey: String? = nil) async throws {
        let normalizedName = capitalizeFirstListLetter(name).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedName.isEmpty else {
            return
        }
        let now = Date().epochMilliseconds
        let localListID = LOCAL_LIST_PREFIX + UUID().uuidString.lowercased()
        let mutationID = UUID().uuidString
        _ = try await cacheManager.updateOfflineState { state in
            var nextState = state
            nextState.lists.append(
                CachedListRecord(
                    id: localListID,
                    name: normalizedName,
                    color: color,
                    iconKey: iconKey,
                    todoCount: 0,
                    updatedAtEpochMs: now,
                    createdAtEpochMs: now
                )
            )
            nextState.pendingMutations.append(
                PendingMutationRecord(
                    mutationId: mutationID,
                    kind: .createList,
                    targetId: localListID,
                    timestampEpochMs: now,
                    title: nil,
                    description: nil,
                    priority: nil,
                    dueEpochMs: nil,
                    rrule: nil,
                    listId: nil,
                    pinned: nil,
                    completed: nil,
                    instanceDateEpochMs: nil,
                    name: normalizedName,
                    color: color,
                    iconKey: iconKey
                )
            )
            return nextState
        }

        do {
            let response = try await api.createList(
                payload: CreateListRequest(name: normalizedName, color: color, iconKey: iconKey)
            )
            guard let createdList = response.list else {
                return
            }
            let createdAt = parseOptionalDate(createdList.createdAt)?.epochMilliseconds ?? now
            let updatedAt = parseOptionalDate(createdList.updatedAt)?.epochMilliseconds ?? now
            _ = try await cacheManager.updateOfflineState { state in
                var nextState = self.replaceLocalListID(state, localListID: localListID, serverListID: createdList.id)
                let todoCount = nextState.todos.filter { !$0.completed && $0.listId == createdList.id }.count
                nextState.lists = nextState.lists.map { list in
                    guard list.id == createdList.id else {
                        return list
                    }
                    return CachedListRecord(
                        id: createdList.id,
                        name: createdList.name,
                        color: createdList.color,
                        iconKey: createdList.iconKey ?? list.iconKey,
                        todoCount: todoCount,
                        updatedAtEpochMs: updatedAt,
                        createdAtEpochMs: createdAt
                    )
                }
                nextState.pendingMutations.removeAll { $0.mutationId == mutationID }
                return nextState
            }
        } catch {
            // Keep the pending CREATE_LIST mutation so background sync can retry it.
        }
    }

    func updateList(listId: String, name: String, color: String? = nil, iconKey: String? = nil) async throws {
        let normalizedName = capitalizeFirstListLetter(name).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !listId.isEmpty, !normalizedName.isEmpty else {
            return
        }
        let now = Date().epochMilliseconds
        if listId.hasPrefix(LOCAL_LIST_PREFIX) {
            _ = try await cacheManager.updateOfflineState { state in
                var nextState = state
                nextState.lists = state.lists.map { list in
                    guard list.id == listId else {
                        return list
                    }
                    return CachedListRecord(
                        id: list.id,
                        name: normalizedName,
                        color: color ?? list.color,
                        iconKey: iconKey ?? list.iconKey,
                        todoCount: list.todoCount,
                        updatedAtEpochMs: now,
                        createdAtEpochMs: list.createdAtEpochMs
                    )
                }
                nextState.pendingMutations = state.pendingMutations.compactMap { mutation in
                    if mutation.kind == .updateList && mutation.targetId == listId {
                        return nil
                    }
                    guard mutation.kind == .createList, mutation.targetId == listId else {
                        return mutation
                    }
                    return PendingMutationRecord(
                        mutationId: mutation.mutationId,
                        kind: mutation.kind,
                        targetId: mutation.targetId,
                        timestampEpochMs: now,
                        title: mutation.title,
                        description: mutation.description,
                        priority: mutation.priority,
                        dueEpochMs: mutation.dueEpochMs,
                        rrule: mutation.rrule,
                        listId: mutation.listId,
                        pinned: mutation.pinned,
                        completed: mutation.completed,
                        instanceDateEpochMs: mutation.instanceDateEpochMs,
                        name: normalizedName,
                        color: color ?? mutation.color,
                        iconKey: iconKey ?? mutation.iconKey
                    )
                }
                return nextState
            }
            _ = await syncManager.syncCachedData(force: true, replayPendingMutations: true)
            return
        }

        _ = try await cacheManager.updateOfflineState { state in
            var nextState = state
            nextState.lists = state.lists.map { list in
                guard list.id == listId else { return list }
                return CachedListRecord(
                    id: list.id,
                    name: normalizedName,
                    color: color ?? list.color,
                    iconKey: iconKey ?? list.iconKey,
                    todoCount: list.todoCount,
                    updatedAtEpochMs: now,
                    createdAtEpochMs: list.createdAtEpochMs
                )
            }
            nextState.pendingMutations.removeAll { $0.kind == .updateList && $0.targetId == listId }
            nextState.pendingMutations.append(
                PendingMutationRecord(
                    mutationId: UUID().uuidString,
                    kind: .updateList,
                    targetId: listId,
                    timestampEpochMs: now,
                    title: nil,
                    description: nil,
                    priority: nil,
                    dueEpochMs: nil,
                    rrule: nil,
                    listId: nil,
                    pinned: nil,
                    completed: nil,
                    instanceDateEpochMs: nil,
                    name: normalizedName,
                    color: color,
                    iconKey: iconKey
                )
            )
            return nextState
        }
        let result = await syncManager.syncCachedData(force: true, replayPendingMutations: true)
        if case let .failure(error) = result, isLikelyUnrecoverableMutationError(error) {
            throw error
        }
    }

    func deleteList(
        listId: String,
        onOptimisticDelete: () -> Void = {}
    ) async throws {
        let normalizedListID = listId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedListID.isEmpty else {
            return
        }

        let now = Date().epochMilliseconds
        let mutationID = UUID().uuidString
        let isLocalOnly = normalizedListID.hasPrefix(LOCAL_LIST_PREFIX)
        _ = try await cacheManager.updateOfflineState { state in
            var nextState = state
            let deletedTodoIDs = Set(state.todos.filter { $0.listId == normalizedListID }.map(\.canonicalId))

            nextState.lists.removeAll { $0.id == normalizedListID }
            nextState.todos.removeAll { $0.listId == normalizedListID }
            nextState.completedItems.removeAll { completed in
                completed.listId == normalizedListID ||
                    completed.originalTodoId.map { deletedTodoIDs.contains($0) } == true
            }
            nextState.pendingMutations.removeAll { mutation in
                mutation.targetId == normalizedListID ||
                    mutation.listId == normalizedListID ||
                    mutation.targetId.map { deletedTodoIDs.contains($0) } == true
            }
            if !isLocalOnly {
                nextState.pendingMutations.append(
                    PendingMutationRecord(
                        mutationId: mutationID,
                        kind: .deleteList,
                        targetId: normalizedListID,
                        timestampEpochMs: now,
                        title: nil,
                        description: nil,
                        priority: nil,
                        dueEpochMs: nil,
                        rrule: nil,
                        listId: nil,
                        pinned: nil,
                        completed: nil,
                        instanceDateEpochMs: nil,
                        name: nil,
                        color: nil,
                        iconKey: nil
                    )
                )
            }
            return nextState
        }

        onOptimisticDelete()

        let result = await syncManager.syncCachedData(force: true, replayPendingMutations: true)
        if case let .failure(error) = result, isLikelyUnrecoverableMutationError(error) {
            throw error
        }
    }

    private func buildLists(from state: OfflineSyncState) -> [ListSummary] {
        let todoCounts = Dictionary(grouping: state.todos.filter { !$0.completed }, by: { $0.listId })
            .mapValues(\.count)
        return orderListsLikeWeb(state.lists)
            .map { listFromCache($0, todoCountOverride: todoCounts[$0.id] ?? 0) }
    }

    private func replaceLocalListID(_ state: OfflineSyncState, localListID: String, serverListID: String) -> OfflineSyncState {
        OfflineSyncState(
            lastSuccessfulSyncEpochMs: state.lastSuccessfulSyncEpochMs,
            lastSyncAttemptEpochMs: state.lastSyncAttemptEpochMs,
            todos: state.todos.map { todo in
                guard todo.listId == localListID else {
                    return todo
                }
                return CachedTodoRecord(
                    id: todo.id,
                    canonicalId: todo.canonicalId,
                    title: todo.title,
                    description: todo.description,
                    priority: todo.priority,
                    dueEpochMs: todo.dueEpochMs,
                    rrule: todo.rrule,
                    instanceDateEpochMs: todo.instanceDateEpochMs,
                    pinned: todo.pinned,
                    completed: todo.completed,
                    listId: serverListID,
                    updatedAtEpochMs: todo.updatedAtEpochMs
                )
            },
            completedItems: state.completedItems.map { completed in
                guard completed.listId == localListID else {
                    return completed
                }
                return CachedCompletedRecord(
                    id: completed.id,
                    originalTodoId: completed.originalTodoId,
                    title: completed.title,
                    description: completed.description,
                    priority: completed.priority,
                    dueEpochMs: completed.dueEpochMs,
                    completedAtEpochMs: completed.completedAtEpochMs,
                    rrule: completed.rrule,
                    instanceDateEpochMs: completed.instanceDateEpochMs,
                    listId: serverListID,
                    listName: completed.listName,
                    listColor: completed.listColor
                )
            },
            lists: state.lists.map { list in
                guard list.id == localListID else {
                    return list
                }
                return CachedListRecord(
                    id: serverListID,
                    name: list.name,
                    color: list.color,
                    iconKey: list.iconKey,
                    todoCount: list.todoCount,
                    updatedAtEpochMs: list.updatedAtEpochMs,
                    createdAtEpochMs: list.createdAtEpochMs
                )
            },
            pendingMutations: state.pendingMutations.map { mutation in
                PendingMutationRecord(
                    mutationId: mutation.mutationId,
                    kind: mutation.kind,
                    targetId: mutation.targetId == localListID ? serverListID : mutation.targetId,
                    timestampEpochMs: mutation.timestampEpochMs,
                    title: mutation.title,
                    description: mutation.description,
                    priority: mutation.priority,
                    dueEpochMs: mutation.dueEpochMs,
                    rrule: mutation.rrule,
                    listId: mutation.listId == localListID ? serverListID : mutation.listId,
                    pinned: mutation.pinned,
                    completed: mutation.completed,
                    instanceDateEpochMs: mutation.instanceDateEpochMs,
                    name: mutation.name,
                    color: mutation.color,
                    iconKey: mutation.iconKey
                )
            },
            aiSummaryEnabled: state.aiSummaryEnabled
        )
    }

    private func capitalizeFirstListLetter(_ value: String) -> String {
        guard let firstLetterIndex = value.firstIndex(where: { $0.isLetter }) else {
            return value
        }
        let current = value[firstLetterIndex]
        let capitalized = String(current).capitalized
        guard String(current) != capitalized else {
            return value
        }

        var result = value
        result.replaceSubrange(firstLetterIndex...firstLetterIndex, with: capitalized)
        return result
    }
}
